machida/lib/wallaroo/experimental/__init__.py

```diff
@@ -12,15 +12,197 @@
+    C = wallaroo._wallaroo_wrap(func.__name__, func, wallaroo.ConnectorEncoder)
+    return C()
+
+
+class StreamDecoderError(Exception):
+    pass
+
+
+class SourceConnectorConfig(object):
+    def __init__(self, host, port, decoder):
+        self._host = host
+        self._port = port
+        self._decoder = decoder
+
+    def to_tuple(self):
+        return ("connector", self._host, str(self._port), self._decoder)
+
+
+class SinkConnectorConfig(object):
+    def __init__(self, host, port, encoder):
+        self._host = host
+        self._port = port
+        self._encoder = encoder
+
+    def to_tuple(self):
+        return ("connector", self._host, str(self._port), self._encoder)
+
+
+class SourceConnector(object):
+    def __init__(self, args=None, required_params=[], optional_params=[]):
+        params = parse_connector_args(args or sys.argv, required_params, optional_params)
+        wallaroo_app = __import__(params.application)
+        actions = wallaroo_app.application_setup(args or sys.argv)
+        try:
+            (_command, _name, port, encoder, _decoder) = next(
+                action for action in actions
+                if action[0] == "source_connector" and action[1] == params.connector_name)
+        except:
+            print("Unable to find a source connector with the name " + params.connector_name)
+            exit(-1)
+        self.params = params
+        self._encoder = encoder
+        self._host = '127.0.0.1'
+        self._port = port
+        self._conn = None
+
+    def connect(self, host=None, port=None):
+        while True:
+            try:
+                conn = socket.socket()
+                conn.connect( (host or self._host, int(port or self._port)) )
+                self._conn = conn
+                return
+            except socket.error as err:
+                if err.errno == socket.errno.ECONNREFUSED:
+                    time.sleep(1)
+                else:
+                    raise
+
+    def write(self, message):
+        # Future parameters
+        partition = None
+        sequence = None
+        if self._conn == None:
+            raise RuntimeError("Please call connect before writing")
+        payload = self._encoder.encode(message)
+        self._conn.sendall(payload)
+
+
+class SinkConnector(object):
+
+    def __init__(self, args=None, required_params=[], optional_params=[]):
+        params = parse_connector_args(args or sys.argv, required_params, optional_params)
+        wallaroo_app = __import__(params.application)
+        actions = wallaroo_app.application_setup(args or sys.argv)
+        try:
+            (_command, _name, port, _encoder, decoder) = next(
+                action for action in actions
+                if action[0] == "sink_connector" and action[1] == params.connector_name)
+        except:
+            print("Unable to find a sink connector with the name " + params.connector_name)
+            exit(-1)
```

Comment:
I don't like this approach with `exit()` calls.
Doing this prevents the user from being able to properly debug from exception messages...
I think it's better to raise a meaningful exception and let the standard error handling in Python (and then Machida) do the rest. It also works better when using a debugger.

As for what the right error is... I think maybe something like  the following would work?
```python
class WallarooTopologyError(exception):
    pass


class SinkConnectorError(WallarooTopologyError):
    pass


...
raise SinkConnectorError("Unable to find a sink connector with the name {}".format(params.connector_name))
```

Reply:
Good point. The only issue here is that these are run within the connector scripts so running a prebuilt script and getting a stack trace rather than an error is a bit ugly. Perhaps I can add an exception handler to the connector scripts.